generator_v3.js

```diff
@@ -140,87 +147,90 @@ var generator_v3 = function(layout, difficulty, bingoList)
 				}
-				// Check if the goal generated is a catalyst for anything already on the sheet
-				else if (typeof currentSheet[indexes[z]].reactant !== 'undefined' && typeof goalCandidate.catalyst !== 'undefined'
-					&& currentSheet[indexes[z]].reactant.some(r=> goalCandidate.catalyst.includes(r)))
+			}
+			// Check if the goal generated is a catalyst for anything already on the sheet
+			if (typeof goalCandidate.catalyst !== 'undefined')
+			{
+				if (reactants.some(r=> goalCandidate.catalyst.includes(r)))
 				{
-					// If it is get a new goal
-					console.log("catalyst/reactant between: " + goalCandidate.name + " and " + currentSheet[indexes[z]].name);
-					cont = false;
-					break;
+					// If it is, get a new goal
+					console.log("reactants for: " + goalCandidate.name);
+					continue GoalGen;
 				}
-				// Check if the goal generated is a reactant for anything already on the sheet
-				else if (typeof currentSheet[indexes[z]].catalyst !== 'undefined' && typeof goalCandidate.reactant !== 'undefined'
-					&& currentSheet[indexes[z]].catalyst.some(r=> goalCandidate.reactant.includes(r)))
+			}
+			// Check if the goal generated is a reactant for anything already on the sheet
+			if (typeof goalCandidate.reactant !== 'undefined')
+			{
+				// If it does, check to see if it's already on the sheet
+				if (catalysts.some(r=> goalCandidate.reactant.includes(r)))
 				{
-					// If it is get a new goal
-					console.log("reactant/catalyst between: " + goalCandidate.name + " and " + currentSheet[indexes[z]].name);
-					cont = false;
-					break;
-				}
-
-				// Compare all the tags of the current goal with the tags of the current one on the sheet
-				if (goalCandidate.tags != null && currentSheet[indexes[z]].tags != null)
-				{
-					for (var x = 0, lenX = goalCandidate.tags.length; x < lenX; x++)
-					{
-						for (var y = 0, lenY = currentSheet[indexes[z]].tags.length; y < lenY; y++)
-						{
-							// If they're the same tag
-							if (goalCandidate.tags[x].name == currentSheet[indexes[z]].tags[y].name)
-							{
-								// If the current goal has a tag that disallows placement on the same line and they are on the same line
-								if (goalCandidate.tags[x].line == false && isOnSameLine(indexes[i], indexes[z]))
-								{
-									console.log("cannot be on same line: " + goalCandidate.name + " and " + currentSheet[indexes[z]].name);
-									cont = false;
-								}
-								// Otherwise just increase the tag count
-								else
-								{
-									tagCount[x]++;
-								}
-							}
-						}
-					}
+					// If it is, get a new goal
+					console.log("catalyst for: " + goalCandidate.name);
+					continue GoalGen;
 				}
 			}
 
-			// If the tag count is higher than the max that tag should have, don't continue
-			if (cont && goalCandidate.tags != null)
+			// If the goal candidate contains tags that cannot be on the same line as other goals with that tag
+			if (goalCandidate.tags.some(r=> r.line == false))
```

Comment:
This `if`, which checks `.line` of the tags is outside of the `if (goalCandidate.tags != null)` above, which is suspicious.

Reply:
Either both of these `if`s should check `.tags` for nullity or we should assume that `.tags` is not `null` for all goals.

Reply:
Oh, there actually exists such an `if` for `goalCandidate`. It is just below, under comment "// If both goals have tags AND are on the same line". Except that it doesn't make sense to check `goalCandidate.tags` for `null` when `.some(..)` has already been called on it.

Reply:
Fixed in da9b544.